[iOS] Double tap does not result in dblclick event dispatch even when the Window object has a listener

aprotyas opened the pull request:
#### dc2b216b9f0f9e15edeae2f97b3c2149e3e0bbe7
<pre>
[iOS] Double tap does not result in dblclick event dispatch even when the Window object has a listener
<a href="https://bugs.webkit.org/show_bug.cgi?id=301558">https://bugs.webkit.org/show_bug.cgi?id=301558</a>
<a href="https://rdar.apple.com/162546881">rdar://162546881</a>

Reviewed by Ryosuke Niwa.

Currently, we walk up the DOM tree from our hit test point to determine
whether any node requires dblclick dispatch.

However, in the case that _only_ the Window object has a dblclick event
listener, we think there is no node responding to double click events,
because the Window object does not show up in the node traversal logic
in LocalFrame::nodeRespondingToDoubleClickEvent().

In this patch, we address this shortcoming by introducing some special
casing for Window. After checking for an appropriate target node (for
dblclick dispatch), we now apply the same logic for Window.

Test: fast/events/touch/ios/double-tap-for-double-click-on-window.html

* LayoutTests/fast/events/touch/ios/double-tap-for-double-click-on-window-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/double-tap-for-double-click-on-window.html: Added.
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/ios/FrameIOS.mm:
(WebCore::LocalFrame::qualifyingNodeAtViewportLocation):
Drive-by rewrite: nullptr is clearer than 0 at the callsite about what
kind of data we are passing in.

* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:
Now that the InteractionInformationAtPosition structure captures more
than just &quot;can the hit tested node respond to dblclick events&quot;, we
rename the flag to `hitNodeOrWindowHasDoubleClickListener`. This
reflects the expanded scope of the bit.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView gestureRecognizerShouldBegin:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::windowWithDoubleClickEventListener):
(WebKit::WebPage::handleDoubleTapForDoubleClickAtPoint):
(WebKit::WebPage::positionInformation):

Canonical link: <a href="https://commits.webkit.org/302274@main">https://commits.webkit.org/302274@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8047a6929ce02041174168545a896a9ca14e2b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128509 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/777 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39340 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135901 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79944 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/eda23a0b-fc33-4d2b-b3d2-d05481219f4a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130381 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/721 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/650 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97831 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65745 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8822d5c2-9beb-4a46-9d46-190a27a33789) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131457 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115139 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78443 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/dc85956f-6dd0-4f99-b360-7b77aea2c607) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/474 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33246 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79184 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108896 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33729 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138350 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/613 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/582 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106366 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/657 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111479 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106180 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27059 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/515 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30019 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52950 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/670 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63870 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/554 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/615 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/624 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1091,15 +1094,37 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
         handleSyntheticClick(std::nullopt, *nodeRespondingToClick, adjustedPoint, modifiers);
 }
 
+static RefPtr<LocalDOMWindow> windowRespondingToDoubleClickEvent(RefPtr<LocalFrame> frame)
```
It's a bit odd to say "responding" since all we know is that there is an event listener. Maybe it would respond, maybe it's a no-op. I'd call this `windowHasDoubleClickEventListeners` or `windowListensToDoubleClickEvent` instead.

rniwa replied:
oh, this returns `LocalDOMWindow` instead of `bool`. Maybe `windowWithDoubleClickEventListener` or `windowWithEventListenerForDoubleClick` then.

rniwa commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -3900,7 +3925,7 @@ static void animationPositionInformation(WebPage& page, const InteractionInforma
 
     if (request.includeHasDoubleClickHandler)
-        info.nodeAtPositionHasDoubleClickHandler = localMainFrame->nodeRespondingToDoubleClickEvent(request.point, adjustedPoint);
+        info.nodeAtPositionHasDoubleClickHandler = localMainFrame->nodeRespondingToDoubleClickEvent(request.point, adjustedPoint) || windowRespondingToDoubleClickEvent(localMainFrame);
```
`nodeAtPositionHasDoubleClickHandler` is misnomer. `window` is definitely not a node, and having an event handler vs. event listener is two different things. We're checking for the presence of event listeners so we should rename this variable (could be done in a separate PR).

rniwa commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -1091,15 +1094,37 @@ static void dispatchSyntheticMouseMove(LocalFrame& mainFrame, const WebCore::Flo
 {
     FloatPoint adjustedPoint;
     RefPtr localMainFrame = m_page->localMainFrame();
     auto* nodeRespondingToDoubleClick = localMainFrame ? localMainFrame->nodeRespondingToDoubleClickEvent(point, adjustedPoint) : nullptr;
```
This function is also oddly named. It should really be renamed to `nodeHasDoubleClickEventListener` or `nodeListensToDoubleClickEvent` (maybe in a separate PR?)

aprotyas replied:
Yes, the naming for these functions (`[node/window]HasDoubleClickEventListener`) and for the `nodeAtPositionHasDoubleClickHandler` flag are slightly misleading. This patch exacerbated the naming issues by continuing with that pattern.

I think it makes sense to address the prior naming issues together in this PR. I'll update shortly.

aprotyas replied:
It seems like there is a bigger precedence for naming `LocalFrame::nodeRespondingTo...()` methods outside of the scope of this change. I will correct my own changes, but fix the existing issues in a separate patch.

aprotyas committed:
[iOS] Double tap does not result in dblclick event dispatch even when the Window object has a listener
https://bugs.webkit.org/show_bug.cgi?id=301558
rdar://162546881

Reviewed by Ryosuke Niwa.

Currently, we walk up the DOM tree from our hit test point to determine
whether any node requires dblclick dispatch.

However, in the case that _only_ the Window object has a dblclick event
listener, we think there is no node responding to double click events,
because the Window object does not show up in the node traversal logic
in LocalFrame::nodeRespondingToDoubleClickEvent().

In this patch, we address this shortcoming by introducing some special
casing for Window. After checking for an appropriate target node (for
dblclick dispatch), we now apply the same logic for Window.

Test: fast/events/touch/ios/double-tap-for-double-click-on-window.html

* LayoutTests/fast/events/touch/ios/double-tap-for-double-click-on-window-expected.txt: Added.
* LayoutTests/fast/events/touch/ios/double-tap-for-double-click-on-window.html: Added.
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/ios/FrameIOS.mm:
(WebCore::LocalFrame::qualifyingNodeAtViewportLocation):
Drive-by rewrite: nullptr is clearer than 0 at the callsite about what
kind of data we are passing in.

* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:
Now that the InteractionInformationAtPosition structure captures more
than just "can the hit tested node respond to dblclick events", we
rename the flag to `hitNodeOrWindowHasDoubleClickListener`. This
reflects the expanded scope of the bit.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView gestureRecognizerShouldBegin:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::windowWithDoubleClickEventListener):
(WebKit::WebPage::handleDoubleTapForDoubleClickAtPoint):
(WebKit::WebPage::positionInformation):

Canonical link: https://commits.webkit.org/302274@main
